## 第十一章 期约与异步函数

### 1. 期约

#### 1.1 期约基础

ES6 新增一个引用类型 Promise , 可以通过 new 操作符来实例化, 创建 新的期约时要传入`执行器函数`作为参数

1. 期约状态机

   `期约是一个由状态的对象`, 可能处于如下三种状态

   + 待定
   + 兑现
   + 拒绝

   重要的点`期约的状态是私有的,无法被外部检测和修改`, 故期约故意将异步操作封装起来,从而隔绝外部的同步代码

2. 解决值 拒绝理由 及 期约用例

   每个期约只要状态切换到兑现,就会有一个`私有的内部值`, 同样的只要状态切换到拒绝, 就会有一个`私有的内部理由`, 无论是值还是理由, 都是原始值的不可修改的引用

3. 通过执行函数控制期约的状态

   在期约的`执行函数`主要有两项职责 : `初始化期约的异步行为`和`控制状态的最终转换`

   `控制状态的最终转换`是通过两个函数参数实现的

   > ​	注意啦! `执行函数是同步执行的`, 意思就是会先执行完执行函数内部代码才会执行外部代码

4. Promise.resolve()

   期约并不是一开始就必须处于待定状态, 通过调用`Promise.resolve()`静态方法, 可以`实例化一个解决的期约`,这个解决的期约`返回的值`对应着传给`Promise.resolve()`的第一个参数, 使用这个方法, `实际上可以将任何一个值转换为一个期约`

   ```js
   setTimeout(console.log, 0, Promise.resolve())
   // Promise <resolved> : undefined
   ```

   `注意:`如果传入的参数本身就是一个期约, 那么它的行为就类似与一个空包装(就相当于啥也没干), 并且会保留传入的期待状态

   ```js
   let p = new Promise(() => {})
   
   setTimeout(console.log, 0, p) // Promise <pending>
   setTimeout(console.log, 0, Promise.resolve(p))  // Promise <pending>
   ```

5. Promise.reject()

   `Promise.reject()`会实例化一个拒绝的期约并抛出一个异步错误(`这个错误不能通过try/catch捕获,而只能通过拒绝处理程序捕获`), 这个拒绝的期约理由就是传给`Promise.reject()`的第一个参数,这个参数会传给后续的拒绝处理程序

   ```js
   let p = Promise.rejcet(3)
   p.then(null, (e) => setTimeout(console.log, 0, e)) //3
   ```

   如果给它传入一个期约对象, 则这个期约会成为它返回的拒绝期约的理由, 就是还是返回一个拒绝期约, 但是拒绝的理由就是你传入的这个期约

   ```js
   setTimeout(console.log, 0, Promise.reject(Promise.resolve()))
   // Promise <rejected> : Promise <resolved>
   ```

6. 同步/异步执行的二元性

   ```js
   try {
       throw new Error('foo')
   } catch(e) {
       console.log(e) //Error : foo
   }
   
   try {
       Promise.reject(new Error('bar'))
   } catch (e) {
       console.log(e) // uncaught(in Promsie)   
   }
   ```

   第一个是正常的捕获到了错误, 但是第二个没有捕获到错误,第二个里面的确是创建一个拒绝实例,并且抛出了拒绝的理由, 但是`因为它没有通过异步模式捕获错误`所以

#### 1.2 期约的实例方法

**Promise.prototype.then()**

这个方法是用来为`期约实例添加处理程序`的主要方法, 注意`Promise.prototype.then()方法返回一个新的期约实例`, 这个新期约的实例是基于onResovled 处理程序的放回值构建

+ onResolved处理函数的返回值会通过Promise.resolve()包装生成新期约
+ 如果没有提供onResolved处理函数, 则Promise.resolve()就会包装上一个期约解决之后的值
+ 如果没有显示的返回语句, 则Promise.resolve()会包装默认的返回值undefined

```js
let p1 = Promise.resolve('foo')

// 若调用then() 不传处理程序 则原样返回
let p2 = p1.then()
console.log(p2) // Promise <resolved> : foo

// 如果有显示的返回, 则 Promise.resolve() 会包装这个值
let p6 = p1.then(() => 'bar')  //Promise <resolved> : bar

```

onRejected 处理程序也类似不过是用来返回要给 Promise.reject() 包装的





